src/sdam/topology.rs

```diff
@@ -0,0 +1,949 @@
+    pub(crate) async fn handle_application_error(
+        &self,
+        address: ServerAddress,
+        error: Error,
+        phase: HandshakePhase,
+    ) -> bool {
+        self.send_message(UpdateMessage::ApplicationError {
+            address,
+            error,
+            phase,
+        })
+        .await
+    }
+
+    /// Update the topology using the provided server description, returning a bool
+    /// indicating whether the topology changed as a result of the update.
+    pub(crate) async fn update(&self, sd: ServerDescription) -> bool {
+        self.send_message(UpdateMessage::ServerUpdate(Box::new(sd)))
+            .await
+    }
+
+    pub(crate) async fn advance_cluster_time(&self, to: ClusterTime) {
+        self.send_message(UpdateMessage::AdvanceClusterTime(to))
+            .await;
+    }
+
+    /// Update the provided state to contain the given list of hosts, removing any
+    /// existing servers whose addresses aren't present in the list.
+    ///
+    /// This will start server monitors for the newly added servers.
+    pub(crate) async fn sync_hosts(&self, hosts: HashSet<ServerAddress>) {
+        self.send_message(UpdateMessage::SyncHosts(hosts)).await;
+    }
+
+    #[cfg(test)]
+    pub(crate) async fn sync_workers(&self) {
+        self.send_message(UpdateMessage::SyncWorkers).await;
+    }
+}
+
+/// Receiver used to listen for updates to the topology.
+pub(crate) struct TopologyUpdateReceiver {
+    update_receiver: UnboundedReceiver<AcknowledgedMessage<UpdateMessage, bool>>,
+}
+
+impl TopologyUpdateReceiver {
+    pub(crate) async fn recv(&mut self) -> Option<AcknowledgedMessage<UpdateMessage, bool>> {
+        self.update_receiver.recv().await
+    }
+}
+
+/// Struct used to get the latest topology state and monitor the topology for changes.
+#[derive(Debug, Clone)]
+pub(crate) struct TopologyWatcher {
+    receiver: watch::Receiver<TopologyState>,
+}
+
+impl TopologyWatcher {
+    fn channel(initial_state: TopologyState) -> (TopologyWatcher, TopologyPublisher) {
+        let (tx, rx) = watch::channel(initial_state);
+        let watcher = TopologyWatcher { receiver: rx };
+        let publisher = TopologyPublisher { state_sender: tx };
+        (watcher, publisher)
+    }
+
+    /// Whether the topology is still active or if all `Client` instances using it have gone
+    /// out of scope.
+    pub(crate) fn is_alive(&self) -> bool {
+        self.receiver.has_changed().is_ok()
+    }
+
+    /// Get a server description for the server at the provided address.
+    pub(crate) fn server_description(&self, address: &ServerAddress) -> Option<ServerDescription> {
+        self.receiver
+            .borrow()
+            .description
+            .get_server_description(address)
+            .cloned()
+    }
+
+    /// Clone the latest state, marking it as seen.
+    pub(crate) fn clone_latest(&mut self) -> TopologyState {
+        self.receiver.borrow_and_update().clone()
```

Comment:
I don't understand the usage of the receiver's "seen" flag.  It's set in `clone_latest` but not `borrow_latest` or `server_description`?

Reply:
The "seen" flag determines whether the currently observed `TopologtState` would be considered "new" or not for the purposes of `wait_for_update`. If we've called `clone_latest` and then immediately call `wait_for_update`, it'll block. If we haven't, it'll return immediately.

Not setting it in `borrow_latest` and `server_description` was more of an ergonomic choice, since it would require those methods to be `&mut self`.

To make this a bit clearer, I renamed the methods to `observe_latest` (formerly `clone_latest`) and `peek_latest` (formerly `borrow_latest`). Let me know if you think these are more helpful / if there are better ones.

Reply:
Much clearer, thank you :)